File: modules/agent/src/controllers/provenance_controller.py
"""Controller layer for provenance routing to services."""

import logging
from typing import Optional, Any

from fastapi import Request

from lib.decorators import handle_http_exceptions
from serializers.provenance import provenance_to_response
from schemas.provenance import ProvenanceCreate
from services.provenance_service import (
    get_provenance as get_provenance_service,
    create_provenance as create_provenance_service,
)

# Re-export for routes
__all__ = ["ProvenanceCreate"]


@handle_http_exceptions
async def get_provenance(entity_type: str, entity_id: int, field_name: Optional[str] = None) -> dict:
    """Get provenance records for an entity."""
    records = await get_provenance_service(entity_type, entity_id, field_name)
    return {"provenance": [provenance_to_response(r) for r in records]}


@handle_http_exceptions
async def create_provenance(request: Request, data: ProvenanceCreate) -> dict:
    """Create a new provenance record."""
    user_id = request.state.user_id
    record = await create_provenance_service(
        data.entity_type, data.entity_id, data.field_name, data.source,
        user_id, data.source_url, data.confidence, data.raw_value
    )
    return {"provenance": provenance_to_response(record)}

File: modules/agent/internal/controllers/provenance_controller.go
package controllers

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"agent/internal/middleware"
	"agent/internal/schemas"
	"agent/internal/services"
)

// ProvenanceController handles provenance HTTP requests
type ProvenanceController struct {
	provenanceService *services.ProvenanceService
}

// NewProvenanceController creates a new provenance controller
func NewProvenanceController(provenanceService *services.ProvenanceService) *ProvenanceController {
	return &ProvenanceController{provenanceService: provenanceService}
}

// GetProvenance handles GET /provenance/{entity_type}/{entity_id}
func (c *ProvenanceController) GetProvenance(w http.ResponseWriter, r *http.Request) {
	entityType := chi.URLParam(r, "entityType")
	entityIDStr := chi.URLParam(r, "entityID")

	entityID, err := strconv.ParseInt(entityIDStr, 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "invalid entity_id")
		return
	}

	var fieldName *string
	if fn := r.URL.Query().Get("field_name"); fn != "" {
		fieldName = &fn
	}

	records, err := c.provenanceService.GetProvenance(entityType, entityID, fieldName)
	if errors.Is(err, services.ErrInvalidEntityType) {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, records)
}

// CreateProvenance handles POST /provenance
func (c *ProvenanceController) CreateProvenance(w http.ResponseWriter, r *http.Request) {
	var input schemas.ProvenanceCreate
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if input.EntityType == "" || input.FieldName == "" || input.Source == "" {
		writeError(w, http.StatusBadRequest, "entity_type, field_name, and source are required")
		return
	}

	var userID *int64
	if uid, ok := middleware.GetUserID(r.Context()); ok {
		userID = &uid
	}

	record, err := c.provenanceService.CreateProvenance(input, userID)
	if errors.Is(err, services.ErrInvalidEntityType) {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusCreated, record)
}
